pkg/controllers/sync/controller.go

```diff
@@ -1104,58 +1111,120 @@ func (s *SyncController) reconcileClusterForCascadingDeletion(
-			} else if err != nil && !meta.IsNoMatchError(err) && !apierrors.IsNotFound(err) {
-				remainingByGVK[gvk] = fmt.Sprintf("Unknown (failed to list from cluster: %v)", err)
+			client, err := s.getClusterClient(cluster.Name)
+			if err != nil {
+				logger.Error(err, "Failed to get cluster client")
+				syncMap.Store(gvk, "failed to get cluster client")
+				return
 			}
-		}
+
+			// update some cluster objects before deleting them
+			var updateFailedObjs []string
+			for i := range clusterObjs {
+				clusterObj := clusterObjs[i].DeepCopy()
+				keyedLogger := logger.WithValues(
+					"namespace", clusterObj.GetNamespace(),
+					"name", clusterObj.GetName(),
+					"gvk", gvk)
+				namespacedName := fmt.Sprintf("%s/%s", clusterObj.GetNamespace(), clusterObj.GetName())
+
+				// get the corresponding federated object
+				fedObjName := naming.GenerateFederatedObjectName(clusterObj.GetName(), ftc.Name)
+				fedObj, err := s.fedAccessor.FederatedResource(common.QualifiedName{
+					Namespace: clusterObj.GetNamespace(),
+					Name:      fedObjName,
+				})
+				if err != nil {
+					keyedLogger.Error(err, "Failed to get corresponding federated object")
+					updateFailedObjs = append(updateFailedObjs, namespacedName)
+					continue
+				}
+				if fedObj == nil {
+					err := fmt.Errorf("the federated object %s is not found", fedObjName)
+					keyedLogger.Error(err, "Failed to get corresponding federated object")
+					updateFailedObjs = append(updateFailedObjs, namespacedName)
+					continue
+				}
+
+				// remove the finalizer added by admiral
+				needUpdate, err := dispatch.RemoveRetainObjectFinalizer(clusterObj)
+				if err != nil {
+					keyedLogger.Error(err, "Failed to remove retain finalizer of target object in cluster")
+					updateFailedObjs = append(updateFailedObjs, namespacedName)
+					continue
+				}
+
+				// if the cluster object need to be orphaned, only remove the managed label
+				if orphaning.ShouldBeOrphaned(fedObj.Object(), clusterObj) {
+					managedlabel.RemoveManagedLabel(clusterObj)
+					needUpdate = true
+				}
+
+				if needUpdate {
+					if _, err = client.Resource(ftc.GetSourceTypeGVR()).Namespace(clusterObj.GetNamespace()).
+						Update(ctx, clusterObj, metav1.UpdateOptions{}); err != nil {
+						keyedLogger.Error(err, "Failed to update target object in cluster")
+						updateFailedObjs = append(updateFailedObjs, namespacedName)
+					}
+				}
+			}
+
+			// if the update operation fails, do not continue the delete operation,
+			// but return directly to avoid accidentally deleting any orphaned cluster objects.
+			if len(updateFailedObjs) > 0 {
+				syncMap.Store(gvk, fmt.Sprintf("failed to update those cluster objects for %s: %v", gvk, updateFailedObjs))
+				return
+			}
+
+			// if the update operation is successful, all managed cluster objects will be deleted.
+			if err := s.deleteClusterObjects(ctx, ftc, namespaces.UnsortedList(), client, cluster.Name); err != nil {
```

Comment:
Why don't we pass `clusterObject` into this function here? Listing again would be very expensive.

Reply:
The code has been updated.